### ISSUE TITLE
Does SAKT overfit?

### ISSUE THREAD

**Issue by jilljenn** [thosgt/edm_main_algorithms#33]
J'ai essayé avec 1000 traces, batch de 1, 1000 epochs, et on dirait que ça overfit.

**Comment by thosgt** [comment]
Oui c'est probable avec si peu de data, tu as essayé avec plus de traces et moins d'epoch ? Au bout de 10 epochs j'avais déjà une performance max. Tu as accès à des gpus ?
Btw je n'ai pas réussi à faire fonctionner le positional embedding correctement. En tout cas ça n'améliorait pas les prédictions.

**Comment by jilljenn** [comment]
Tu as donc réussi à reproduire les super résultats de Pandey ? Si oui, c'est encourageant :)

**Comment by jilljenn** [comment]
OK au départ j'avais confondu le nombre de traces avec le nombre d'étudiants.
J'ai réussi à configurer les GPU de Colab.

- Tu as remarqué que n_traces = 10000 correspond à en fait seulement 39 étudiants ?
- Du coup le batch_size par défaut de 100 n'a pas beaucoup d'intérêt.
- Vous n'avez pas tronqué les séquences à 100 traces comme préconisé par Pandey, c'est normal ?
- Qu'entends-tu par "performance max" ?

**Comment by jilljenn** [comment]
Avec ces paramètres :


    params = {
      'dataset': 'assistments09',
      'logdir': 'runs/sakt',
      'embed_inputs': 'store_true',
      'embed_size': 100,
      'hid_size': 100,
      'num_heads': 5,
      'encode_pos': True,
      'drop_prob': 0.2,
      'batch_size': 100,
      'lr': 1e-3,
      'num_epochs': 100,
      'n_traces': 0,
      'disable_cuda': False,
      'device': torch.device('cuda')
    }

J'obtiens péniblement 0.699 de AUC sur le validation set après 60 epochs. L'AUC de train plafonne à 0.770, ce qui est étrange.

**Comment by thosgt** [comment]
Tu as testé sans positional encoding ? Ça faisait diminuer mes résultats.
Le n_traces faible était pour tester que l'algo tournait. Mais je peux l'enlever si ça apporte de la confusion.

**Comment by thosgt** [comment]
Et non je n'ai pas réussi à reproduire les résultats. 
Je n'avais pas vu qu'ils recommandaient de tronquer les séquences en effet.

**Comment by Ticus0228** [comment]
I have a problem with the assistments2009 dataset. 
I tried your SAKT and did some modification to it (likes freeze position embedding or not), but the best test AUC is around 70%. 
Do you get the test AUC=84.8% in assistments2009 as mentioned in SAKT paper? 
I am looking forward to your reply~ ^v^

**Comment by thosgt** [comment]
Hi Ticus,

No I didn't manage to get the test AUC as high as in the paper, not even close. They use learned embeddings contrary to me but I doubt it made them gain that much AUC.
Let me know if you manage to improve the model or if you found flaws in my code.

Best,

Thomas

**Comment by Ticus0228** [comment]
Hi Thomas, 

Thank you for your reply~
I will inform you if I improve your model.

Best, 

Ticus
